Common/src/geometry/CGeometry.cpp

```diff
@@ -3637,7 +3638,56 @@ const CCompressedSparsePatternUL& CGeometry::GetEdgeColoring(su2double* efficien
 
     /*--- Color the edges. ---*/
     constexpr bool balanceColors = true;
-    edgeColoring = colorSparsePattern(pattern, edgeColorGroupSize, balanceColors);
+
+    /*--- Lambda to account for different numbers of colors. ---*/
+    auto getColorSparsePattern = [&](unsigned long edgeColorGroupSize) {
+      if (largeNumberOfColors) {
+        return colorSparsePattern<unsigned char, 255>(pattern, edgeColorGroupSize, balanceColors);
+      } else {
+        return colorSparsePattern(pattern, edgeColorGroupSize, balanceColors);
+      }
+    };
+
+    /*--- if requested, find an efficient coloring with maximum color group size (up to edgeColorGroupSize) ---*/
+    if (maximizeEdgeColorGroupSize) {
+      auto upperEdgeColorGroupSize = edgeColorGroupSize + 1; /* upper bound that is deemed too large */
+      auto nextEdgeColorGroupSize = edgeColorGroupSize;      /* next value that we are going to try */
+      auto lowerEdgeColorGroupSize = 1ul;                    /* lower bound that is known to work */
+
+      while (true) {
+        auto currentEdgeColoring = getColorSparsePattern(nextEdgeColorGroupSize);
+
+        /*--- if the coloring fails, reduce the color group size ---*/
+        if (currentEdgeColoring.empty()) {
+          upperEdgeColorGroupSize = nextEdgeColorGroupSize;
+          nextEdgeColorGroupSize = lowerEdgeColorGroupSize + (upperEdgeColorGroupSize - lowerEdgeColorGroupSize) / 2;
+          continue;
+        }
+
+        const su2double currentEfficiency =
+            coloringEfficiency(currentEdgeColoring, omp_get_max_threads(), nextEdgeColorGroupSize);
+
+        /*--- if the coloring is not efficient, reduce the color group size ---*/
+        if (currentEfficiency < COLORING_EFF_THRESH) {
+          upperEdgeColorGroupSize = nextEdgeColorGroupSize;
+        }
+        /*--- otherwise, try to enlarge the color group size ---*/
+        else {
+          lowerEdgeColorGroupSize = nextEdgeColorGroupSize;
+        }
+        auto increment = (upperEdgeColorGroupSize - lowerEdgeColorGroupSize) / 2;
+
+        nextEdgeColorGroupSize = lowerEdgeColorGroupSize + increment;
+
+        if (increment == 0) {
+          break;
+        }
+      }
+
+      edgeColorGroupSize = nextEdgeColorGroupSize;
+    }
+
+    edgeColoring = getColorSparsePattern(edgeColorGroupSize);
```

Comment:
It could be rewritten so that the last coloring evaluated in the loop is immediately used as the final coloring. So yeah, the final coloring is computed twice with adaptivity at the moment, I'll change it.

Reply:
The last evaluated coloring is not necessarily identical to the final coloring. If the last coloring is not admissible, the final coloring still needs to be recomputed.

Reply:
Makes sense :+1: